GVFS/GVFS.UnitTests/Common/OrgInfoApiClientTests.cs

```diff
@@ -0,0 +1,110 @@
+            new OrgInfo() { OrgName = "org1", Platform = "macOS", Ring = "fast", Version = "1.2.3.3" },
+            new OrgInfo() { OrgName = "org1", Platform = "macOS", Ring = "slow", Version = "1.2.3.4" },
+            new OrgInfo() { OrgName = "org2", Platform = "windows", Ring = "fast", Version = "1.2.3.5" },
+            new OrgInfo() { OrgName = "org2", Platform = "windows", Ring = "slow", Version = "1.2.3.6" },
+            new OrgInfo() { OrgName = "org2", Platform = "macOS", Ring = "fast", Version = "1.2.3.7" },
+            new OrgInfo() { OrgName = "org2", Platform = "macOS", Ring = "slow", Version = "1.2.3.8" },
+        };
+
+        private string baseUrl = "https://www.contoso.com";
+
+        private interface IHttpMessageHandlerProtectedMembers
+        {
+            Task<HttpResponseMessage>  SendAsync(HttpRequestMessage message, CancellationToken token);
+        }
+
+        [TestCaseSource("TestOrgInfo")]
+        public void QueryNewestVersionWithParams(OrgInfo orgInfo)
+        {
+            Mock<HttpMessageHandler> handlerMock = new Mock<HttpMessageHandler>(MockBehavior.Strict);
+
+            handlerMock.Protected().As<IHttpMessageHandlerProtectedMembers>()
+                .Setup(m => m.SendAsync(It.Is<HttpRequestMessage>(request => this.UriMatches(request.RequestUri, this.baseUrl, orgInfo.OrgName, orgInfo.Platform, orgInfo.Ring)), It.IsAny<CancellationToken>()))
+                .ReturnsAsync(new HttpResponseMessage()
+                {
+                    StatusCode = HttpStatusCode.OK,
+                    Content = new StringContent(this.ConstructResponseContent(orgInfo.Version))
+                });
+
+            HttpClient httpClient = new HttpClient(handlerMock.Object);
+
+            OrgInfoApiClient upgradeChecker = new OrgInfoApiClient(httpClient, this.baseUrl);
+            Version version = upgradeChecker.QueryNewestVersion(orgInfo.OrgName, orgInfo.Platform, orgInfo.Ring);
+
+            version.ShouldEqual(new Version(orgInfo.Version));
+
+            handlerMock.VerifyAll();
+        }
+
+        private bool UriMatches(Uri uri, string baseUrl, string expectedOrgName, string expectedPlatform, string expectedRing)
+        {
+            bool hostMatches = uri.Host.Equals(baseUrl);
+
+            Dictionary<string, string> queryParams = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            foreach (string param in uri.Query.Substring(1).Split('&'))
+            {
+                string[] fields = param.Split('=');
+                string key = fields[0];
+                string value = fields[1];
+
+                queryParams.Add(key, value);
+            }
+
+            if (queryParams.Count != 3)
+            {
+                return false;
+            }
+
+            if (!queryParams.TryGetValue("Organization", out string orgName) || !string.Equals(orgName, expectedOrgName, StringComparison.OrdinalIgnoreCase))

```

Comment:
Can this just use a `ShouldEqual` and have the test fail?

Reply:
This is used in determining whether a Moq setup matches a call. If the setup expectations do not match, then the test will fail elsewhere.

Reply:
But will we know elsewhere that it was because the orgName didn't match the expectedOrgName or will it be some obscure failure that will be hard to determine what failed?

Reply:
We will know that the method was not called with the expected parameters, which I think is pretty close. In my experience, it also prints out the parameters from the actual calls (although, I am not sure exactly it prints out in this case).